core/templates/dev/head/pages/exploration_editor/ExplorationStatesServiceSpec.js

```diff
@@ -0,0 +1,175 @@
+// Copyright 2018 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Tests for ExplorationStatesService.
+ */
+
+describe('ExplorationStatesService', function() {
+  beforeEach(module('oppia'));
+
+  beforeEach(inject(function($injector) {
+    this.ess = $injector.get('ExplorationStatesService');
+
+    spyOn($injector.get('ExplorationContextService'), 'getExplorationId')
+      .and.returnValue('7');
+  }));
+
+  describe('Callback Registration', function() {
+    beforeEach(function() {
+      this.ess.init({
+        Hola: {
+          content: '',
+          param_changes: [],
+          interaction: {
+            answer_groups: [{
+              rule_specs: [{rule_type: 'Contains', inputs: {x: 'hola'}}],
+              outcome: {
+                dest: 'Me Llamo',
+                feedback: {html: 'buen trabajo!'},
+                labelled_as_correct: true
+              }
+            }],
+            default_outcome: {
+              dest: 'Hola',
+              feedback: {html: 'try again!'},
+              labelled_as_correct: false
+            },
+            hints: [],
+            id: 'TextInput',
+          },
+          classifier_model_id: 0,
+          content_ids_to_audio_translations: {}
+        }
+      });
+    });
+    beforeEach(inject(function($injector) {
+      // ChangeListService will need its calls mocked out since it isn't
+      // configured correctly in, or interesting to, the tests of this block.
+      this.cls = $injector.get('ChangeListService');
+    }));
+
+    describe('.registerOnStateAddedCallback', function() {
+      beforeEach(function() {
+        spyOn(this.cls, 'addState');
+      });
+
+      it('callsback when a new state is added', function() {
+        var callbackSpy = jasmine.createSpy('callback');
+
+        this.ess.registerOnStateAddedCallback(callbackSpy);
+        this.ess.addState('Me Llamo');
+
+        expect(callbackSpy).toHaveBeenCalledWith('Me Llamo');
+      });
+
+      it('does not accept duplicate callbacks', function() {
```

Comment:
See above -- I think duplicate callbacks and having the thing do a simple append() is OK. Unless we're concerned with the possibility of multiple registration calls? If so, consider either the first alternative I mentioned in the comment above (if we never need more than one callback), or alternatively have a frontend Karma test that loads the page more or less in its entirety and verifies how many callbacks are made to addState in "normal operation" (e.g. even if multiple addState operations happen, there should still only be a single callback in the arrays).

Reply:
Hmm, yep that's exactly why I was concerned (possibility of multiple registration calls). I see, yes that's a good point I'll refactor it so that the STAS is responsible for ensuring that it doesn't register more than once.

Reply:
Done.

Reply:
Done.